Stabilize 5 unstable tests

alirezazamani opened the pull request:
### Issues
- [x] My PR addresses the following Helix issues and references them in the PR title:
Fixes #619 , #620 , #621 , #622 , #623 .

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
In this commit three test that were unstable for some time has been stabilized.
1- TestWorkflowControllerDataProvider:
This test were rely on thread sleep. This has been changed to TestHelper.verify.
2-TestDrop:
This test were rely on thread sleep. This has been changed to TestHelper.verify.
3-TestGetLastScheduledTaskExecInfo:
This test were rely on thread sleep. This has been changed to TestHelper.verify.
Also a check has been added to measure the scheduled tasks and compared the
result with the expected number of tasks that should be scheduled.
Also in this test some of the hard coded strings for queue naming
has been removed and replaced with TestHelper.getTestMethodName().
4- TestStopWorkflow:
At the beginning of each test, participant have started and at the end of the
test, participants have stopped. Hence dependency between the tests added to
avoid potential conflicts between individual tests. Also one global variable
has been use among the task to stop which has been changed to local variable.
Also Thread.sleep has been eliminated.
5- TestClusterVerifier:
This test were rely on thread sleep. This has been changed to TestHelper.verify.

### Tests

- [x] The following is the result of the "mvn test" command on the appropriate module:

Test Result 1: mvn test
[INFO] Tests run: 886, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3,286.007 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 886, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  54:51 min
[INFO] Finished at: 2019-11-20T10:10:53-08:00
[INFO] ------------------------------------------------------------------------


### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- [x] My diff has been formatted using helix-style.xml

narendly commented on helix-core/src/test/java/org/apache/helix/controller/dataproviders/TestWorkflowControllerDataProvider.java:
```diff
@@ -30,21 +30,26 @@
+      int workflowConfigMapSize = cache.getWorkflowConfigMap().size();
+      int contextsSize = cache.getContexts().size();
+      return (configMapSize == 1 && workflowConfigMapSize == 1 && contextsSize == 2);
+    }, 30 * 1000);
```
Is there any significance to the wait times (30 * 1000 or 10 * 1000's)? If not, could we make it a public static constant in TestHelper and use that throughout these tests?

alirezazamani replied:
Good comment. Fixed. We already have TestHelper.WAIT_DURATION. I leveraged that here.

narendly commented on helix-core/src/test/java/org/apache/helix/integration/task/TestStopWorkflow.java:
```diff
@@ -203,13 +202,20 @@ public void testResumeTaskForQuota() throws InterruptedException {
+    // Check the jobs are in progress. Each job has one task.
+    for (int i = 0; i < 30; i++) {
+      _driver.pollForJobState(workflowName_1, workflowName_1 + "_JOB" + i, TaskState.IN_PROGRESS);
+    }
```
Nit: Could a job be in progress but its task not in RUNINNG state? Do you need to check for the task state?

alirezazamani replied:
Yes. It can also be in INIT. I added the check got the tasks.

narendly commented on helix-core/src/test/java/org/apache/helix/integration/task/TestStopWorkflow.java:
```diff
@@ -275,13 +281,16 @@ private void cleanupParticipants(int numNodes) {
    * A mock task class that models a short-lived task to be stopped.
    */
   private class StopTask extends MockTask {
+    boolean toStop = false;
```
What does "toStop" mean? Is there a better name for this?

alirezazamani replied:
Fixed. I changed it to _stopFlag.

narendly commented on helix-core/src/test/java/org/apache/helix/task/TestGetLastScheduledTaskExecInfo.java:
```diff
@@ -41,37 +42,47 @@ public void beforeClass() throws Exception {
+  public void testGetLastScheduledTaskExecInfo() throws Exception {
+    // Start new queue that has one job with long tasks and record start time of the tasks
+    String queueName = TestHelper.getTestMethodName();
+    List<Long> startTimesWithStuckTasks = setupTasks(queueName, 5, 99999999, 2);
```
Nit: could we make 99999999 a constant and name it something more appropriate?

alirezazamani replied:
Addressed.

narendly commented on helix-core/src/test/java/org/apache/helix/task/TestGetLastScheduledTaskExecInfo.java:
```diff
@@ -41,37 +42,47 @@ public void beforeClass() throws Exception {
-    execInfo = _driver.getLastScheduledTaskExecutionInfo("TestWorkflow_3");
+    // Stop and delete the queue
+    _driver.stop(queueName);
+    _driver.deleteAndWaitForCompletion(queueName, 30 * 1000);
```
Do you think you could make recurring 30 * 1000's a constant?

alirezazamani replied:
Fixed.

narendly commented on helix-core/src/test/java/org/apache/helix/task/TestGetLastScheduledTaskExecInfo.java:
```diff
@@ -41,37 +42,47 @@ public void beforeClass() throws Exception {
+          && execInfo.getTaskPartitionState() == TaskPartitionState.COMPLETED
+          && execInfo.getStartTimeStamp().equals(lastScheduledTaskTs)
+          && startTimesFastTasks.get(startTimesFastTasks.size() - 1).equals(lastScheduledTaskTs));
+    }, 30 * 1000);
```
magic number

alirezazamani replied:
Fixed.

narendly commented on helix-core/src/test/java/org/apache/helix/task/TestGetLastScheduledTaskExecInfo.java:
```diff
@@ -41,37 +42,47 @@ public void beforeClass() throws Exception {
-    Assert.assertEquals(execInfo.getStartTimeStamp(), lastScheduledTaskTs);
+    // Start the new queue with new task configuration.
+    // Start new queue that has one job with 4 short tasks and record start time of the tasks
+    List<Long> startTimesFastTasks = setupTasks(queueName, 4, 10, 4);
```
Explain what these numbers mean. 

10 -> make it a constant?

alirezazamani replied:
Fixed. Thanks for the commend.

jiajunwang commented on helix-core/src/test/java/org/apache/helix/tools/TestClusterVerifier.java:
```diff
@@ -132,8 +132,14 @@ public void testDisablePartitionAndStopInstance() throws InterruptedException {
         Lists.newArrayList(FULL_AUTO_RESOURCES[0] + "_0"));
-    Thread.sleep(1000);
-    Assert.assertFalse(strictMatchVerifier.verify(3000));
+
```
Please refer to my change in the wagedRebalancer branch, using the bestPossibleVerifier directly should help to solve the problem.
Although I'm not sure if that is also related to what I have changed, could you please try that one?

alirezazamani replied:
Thanks for the comment. I removed the Thread.sleep and change it with BestPossibleExternalViewVerifier to make sure resources have converged. As we discussed, I also add BestPossibleExternalViewVerifier next to StrictMatchExternalViewVerifier to make sure test functionality has not been changed. Thanks.

jiajunwang commented on helix-core/src/test/java/org/apache/helix/tools/TestClusterVerifier.java:
```diff
@@ -173,15 +179,15 @@ public void testResourceSubset() throws InterruptedException {
             .setResources(Sets.newHashSet(testDB)).build();
     Assert.assertTrue(verifier.verifyByPolling());
 
+    _admin.enableCluster(_clusterName, false);
```
This changes the purpose of this test.
The BestPossibleExternalViewVerifier and StrictMatchExternalViewVerifier should be tested with exactly the same condition.

Why do we need to move these 2 changes to make BestPossibleExternalViewVerifier verifying pass?

alirezazamani replied:
Fixed. Thanks. Now those checks are together.

alirezazamani commented:
@dasahcc  I did put some general information about the fixes in the PR description.
Here are more detail:

1- TestWorkflowControllerDataProvider:
In this test, the Tread.sleep(4000) has been used followed by multiple Assert statements to check values such as cache.getJobConfigMap().size() and cache.getWorkflowConfigMap().size(). However, Thread.sleep() is not desirable since there might be fast or slow controller threads that reach the expected values sooner/later than expected. Hence here it is preferable to used TestHelper.verify to check read these values and check them in a loop. 

2- TestDrop: In this test Thread.sleep has been used followed by Assert which is eliminated in this PR and has been substituted with TestHelper.verify.

3- TestGetLastScheduledTaskExecInfo:
This test were rely on Thread.sleep(). This has been changed to TestHelper.verify.
In this test, one second delay has been used to make sure controller has enough time to schedule the tasks. However, this one second is not enough in most of the cases. Hence, in this PR, instead of Thread.sleep, it is better and more robust to use TestHelper.verify and count the number of scheduled task by checking their start time. If number of tasks with start time reaches the expected number of scheduled tasks, it means the test can continue to the next steps.

4- TestStopWorkflow:
In this test Thread.sleep has been used to make sure all of the tasks are running. Instead of Thread.sleep, I used pollForJobState to make sure jobs are running and I also got all the tasks status in the context and checked if the tasks have reached the running state. Hence, we can be sure that the queue are in desired state after these checks. Also, previously, a global variable has been use among all the tasks to stop the tasks. Having a global variable to stop all the task might cause race condition between the tasks (concurrent get and set with different order). This has been changed to local variable.

alirezazamani commented:
Result of mvn test after the addressing the comments:
[INFO] Tests run: 886, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3,262.438 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 886, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  54:27 min
[INFO] Finished at: 2019-11-25T16:08:16-08:00
[INFO] ------------------------------------------------------------------------

narendly commented on helix-core/src/test/java/org/apache/helix/task/TestGetLastScheduledTaskExecInfo.java:
```diff
@@ -29,10 +29,14 @@
   private final static long INVALID_TIMESTAMP = -1L;
+  private static int SHORT_EXECUTION_TIME = 10;
+  private static int LONG_EXECUTION_TIME = 99999999;
+  private static final int DELETE_DELAY = 30 * 1000;
```
nit: make them longs with L appended at the end :)

alirezazamani replied:
Fixed. Thanks.

narendly commented on helix-core/src/test/java/org/apache/helix/tools/TestClusterVerifier.java:
```diff
@@ -132,8 +132,14 @@ public void testDisablePartitionAndStopInstance() throws InterruptedException {
+          new StrictMatchExternalViewVerifier.Builder(_clusterName).setZkClient(_gZkClient).build();
+      boolean verified = strictMatchVerifierTemp.verify(3000);
+      return (!verified);
+    }, 60 * 1000);
```
Magic number?

alirezazamani replied:
Good catch. Fixed.

narendly commented on helix-core/src/test/java/org/apache/helix/integration/TestDrop.java:
```diff
@@ -55,16 +55,26 @@
+      }, TestHelper.WAIT_DURATION);
+      Assert.assertTrue(isCurrentStateNull);
+
+
```
Redundant space?

alirezazamani replied:
Fixed.

narendly commented on helix-core/src/test/java/org/apache/helix/controller/dataproviders/TestWorkflowControllerDataProvider.java:
```diff
@@ -30,21 +30,26 @@
+      return (configMapSize == 1 && workflowConfigMapSize == 1 && contextsSize == 2);
+    }, TestHelper.WAIT_DURATION);
+    Assert.assertTrue(expectedValuesAchieved);
 
```
Redundant space?

alirezazamani commented:
This PR is ready to be merged, approved by @narendly.

Final commit message:

Title: Stabilize 5 unstable tests

Body:
This commit, stabilizes five tests that have been unstable for some time.
These tests are:
1- TestWorkflowControllerDataProvider.
2-TestDrop
3-TestGetLastScheduledTaskExecInfo
4- TestStopWorkflow
5- TestClusterVerifier:
